quickwit/quickwit-ingest/src/ingest_v2/debouncer.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use quickwit_proto::control_plane::{
+    GetOrCreateOpenShardsRequest, GetOrCreateOpenShardsSubrequest,
+};
+use quickwit_proto::ingest::ShardIds;
+use quickwit_proto::types::{IndexId, SourceId};
+use tokio::sync::{OwnedRwLockWriteGuard, RwLock};
+
+/// Debounces [`GetOrCreateOpenShardsRequest`](quickwit_proto::control_plane::GetOrCreateOpenShardsRequest) requests by index and source IDs. It gives away a permit to the first request and a barrier to subsequent requests.
+#[derive(Default)]
+pub(super) struct Debouncer {
+    locks: HashMap<(IndexId, SourceId), Arc<RwLock<()>>>,
+}
+
+impl Debouncer {
+    pub async fn acquire(
+        &mut self,
+        index_id: &str,
+        source_id: &str,
+    ) -> Result<PermitGuard, BarrierGuard> {
+        let key = (index_id.to_string(), source_id.to_string());
+        let lock = self.locks.entry(key).or_default();
+
+        if let Ok(permit) = lock.clone().try_write_owned() {
+            Ok(PermitGuard(permit))
+        } else {
+            let barrier = lock.clone();
+            Err(BarrierGuard(barrier))
+        }
+    }
+}
+
+pub(super) struct PermitGuard(OwnedRwLockWriteGuard<()>);
+
+pub(super) struct BarrierGuard(Arc<RwLock<()>>);
+
+impl BarrierGuard {
+    pub async fn wait(self) {
+        let _ = self.0.read().await;
+    }
+}
+
+#[derive(Default)]
+pub(super) struct Rendezvous {
+    permits: Vec<PermitGuard>,
+    barriers: Vec<BarrierGuard>,
+}
+
+impl Rendezvous {
+    /// Releases the permits and waits for the barriers to be released.
+    pub async fn wait(mut self) {
+        self.permits.clear();
```

Comment:
```suggestion
        // Releasing the permits before waiting for barrier is necessary to avoid
        // dead locks.
        self.permits.clear();
```